File: DPCodes/100-647-countSubStrings.cpp
#include<string>
#include<iostream>

using STR=std::string;
using std::cout;


class Solution {
public:
    int countSubstrings(STR s) {
        int num = 0;
        int n = s.size(); 
        for(int i=0;i<n;i++)//遍历回文中心点
        {
            for(int j=0;j<=1;j++)//j=0,中心是一个点，j=1,中心是两个点
            {
                int l = i;
                int r = i+j;
                while(l>=0 && r<n && s[l--]==s[r++])num++;
            }
        }
        return num;
    }
    int  spread(const STR& s,int l,int r)
    {
        int count=0;
        int sz=s.size();
        while(l>=0&&r<sz&&s[l]==s[r])
        {
            count+=1;
            l-=1;
            r+=1;
        }
        return count;
    }
    int countSubstrings2(const STR& s) {
        int ans=0;
        int n=s.size();
        for(int i=0;i<n;++i)
        {
            ans+=spread(s,i,i);
        }
        for(int i=0;i<n-1;++i)
        {
            ans+=spread(s,i,i+1);
        }
        return ans;
    }
};

int main()
{
    STR S{"aaa"};
    cout<<Solution().countSubstrings2(S);
}

File: DPCodes/100-647-countSubStrings.py
class Solution:
    def countSubstrings(self, s):
        '''
        Find number of palindromic strings in s. bruteForce method
        Args:
            s (str): String to analyze
        Returns:
            int: Count of palindromes
        '''
        palindromes = 0
        for i in range(len(s)):
            for j in range(i, len(s)):
                if s[i:j] == s[j:i:-1]:
                    palindromes += 1
        return palindromes
    
    def countSubstrings2(self,s):
        '''
        This solution builds up palindromes one character at a time on each end of the substring. 
        It has to run two while loops for each character in the string s:
               one for odd length palindromes and one for even length palindromes.
        '''
        ans=0
        for i in range(len(s)):
            left,right=i,i
            while(left>=0 and right<len(s) and s[left]==s[right]):
                ans+=1
                left-=1
                right+=1
            
            left,right=i,i+1
            while(left>=0 and right<len(s) and s[left]==s[right]):
                ans+=1
                left-=1
                right+=1
                
        return ans
    
s="aaa"
print(Solution().countSubstrings(s))
print(Solution().countSubstrings2(s))

    
    